lalrpop/src/lr1/interpreter.rs

```diff
@@ -0,0 +1,273 @@
+    start_symbol: NonterminalString,
+
+    /// the vector of states
+    states: &'emitter [State<'grammar>],
+
+    /// where we write output
+    out: &'emitter mut RustWrite<Vec<u8>>,
+
+    nonterminal_bits: Map<NonterminalString, usize>,
+
+}
+
+pub fn compile<'emitter, 'grammar>(grammar: &'grammar Grammar,
+                                   user_start_symbol: NonterminalString,
+                                   start_symbol: NonterminalString,
+                                   states: &'emitter [State<'grammar>],
+                                   out: &'emitter mut RustWrite<Vec<u8>>)
+                                   -> io::Result<()> {
+    let mut interpreter = Interpreter::new(grammar, user_start_symbol, start_symbol, states, out);
+
+    interpreter.write()
+}
+
+
+impl<'emitter, 'grammar> Interpreter<'emitter, 'grammar> {
+    fn new(grammar: &'grammar Grammar,
+               user_start_symbol: NonterminalString,
+               start_symbol: NonterminalString,
+               states: &'emitter [State<'grammar>],
+               out: &'emitter mut RustWrite<Vec<u8>>)
+               -> Interpreter<'emitter, 'grammar>
+    {
+        Interpreter {
+            grammar: grammar,
+            prefix: &grammar.prefix,
+            types: &grammar.types,
+            user_start_symbol: user_start_symbol,
+            start_symbol: start_symbol,
+            states: states,
+            out: out,
+            nonterminal_bits: grammar.nonterminals.keys().cloned().zip(0..).collect(),
+        }
+    }
+
+    fn parse_error_type(&mut self) -> String {
+        format!("{}ParseError<{},{},{}>",
+                self.prefix,
+                self.types.terminal_loc_type(),
+                self.types.terminal_token_type(),
+                self.types.error_type())
+    }
+
+    fn write_action_table(&mut self) -> io::Result<()> {
+        let row_len = self.grammar.all_terminals.len()+1;
+
+        let productions = self.grammar.nonterminals.values().flat_map(|nt| nt.productions.iter()).cloned().collect::<Vec<_>>();
+        let production_bits: Map<_, _> = productions.iter().cloned().zip(0..).collect();
+
+        rust!(self.out, "const productions: [ReducedProduction; {}] = [", productions.len());
+
+        for p in &productions {
+            rust!(self.out, "    ReducedProduction {{ nonterminal: {}, symbol_count: {} }},", *self.nonterminal_bits.get(&p.nonterminal).expect("got nonexisting nonterminal"), p.symbols.len());
+        }
+        rust!(self.out, "];");
+
+        let rows = try!(self.states.iter().enumerate().map(|(i, s)| {
+            let mut row = iter::repeat("0".to_owned()).take(row_len).collect::<Vec<_>>();
+            for (lookahead, action) in &s.tokens {
+                let target = match *action {
+                    Action::Shift(index) => format!("{}", (index.0 as i32) + 1),
+                    Action::Reduce(ref prod) => format!("-{}", production_bits.get(prod).expect("got nonexisting production")),
```

Comment:
Nit: I personally prefer to use `[]` than `get/expect`. The latter doesn't give me any more information than a backtrace -- and, in fact, if we ever improve the error message for `[]` (which we can, with specialization!), would give me *less* since it wouldn't tell me the value of the key.